src/lib/url.js

```diff
@@ -53,10 +56,64 @@ export const parseQueryParameterFamily = (baseName, query) => {
 export const parseDictionaryParameter = (baseName, query) => {
   const family = parseQueryParameterFamily(baseName, query);
   return Object.keys(family).reduce((dictionary, key) => {
-    return Object.assign({}, dictionary, {[key]: new Set(parseListParameter(family[key]))});
+    return Object.assign({}, dictionary, {
+      [key]: new Set(parseListParameter(family[key])),
+    });
   }, {});
 };
 
+export const compileListParameter = value => {
+  return [...value].join(',');
+};
+
+export const compileDictionaryParameter = (baseName, type, query) => {
+  return `${baseName}[${type}]=${compileListParameter(query[type])}`;
+};
+
+export const compileQueryParameterFamily = (baseName, query) => {
+  const merge = (a, b) => {
+    return `${a}=${encodeURIComponent(b)}`;
+  };
+
+  const compile = (current, param) => {
+    if (typeof param === 'object') {
+      for (let [key, value] of Object.entries(param)) {
+        current += `[${key}]`;
+        if (value instanceof Set) {
+          return [...value].map(val => `${current}[]=${val}`).join('&');
+        } else if (typeof value === 'string') {
+          return merge(current, value);
+        } else if (typeof value === 'object') {
+          return Object.keys(value)
+            .map(nextKey => compile(current, { [nextKey]: value[nextKey] }))
+            .join('&');
```

Comment:
This extra map seems unnecessary. I feel like you should be able to call `compile` with `value` directly. I'm sure you tried that and it didn't work. Why though? My suspicion: because you're doing `.join('&')` within the recursive function.

I think this would be easier to reason about if it first created a 1-level deep object before turning it into a string: 

```
const normalized = {
   "foo[bar][baz]" = [..., ...],
   "bar[baz][]" = ...,
   "bar[baz][]" = ...,
};

const serialized = normalized.entries().reduce((queryString, [key, value]) => {
  return `${queryString}&` + `${key}=${Array.isArray(value) ? value.join(',') : value}`;
});
```

Reply:
A reduce might be better, I didn't like how I had to pass the iterating key back into a new object. The reason I used a map with a join is because of cases like
`{ foo: { bar: 'qux', baz: 'quux' } }`

Where the "value" had multiple keys (bar, baz) and needed to loop over those while joining the whole current + additional. 

`'filter[foo][bar]=qux&filter[foo][baz]=quux'`

Reply:
Ok I rewrote it a bit, lmk what you think about this version instead:

```js
export const compileQueryParameterFamily2 = (baseName, query) => {
  const create = (path, value) => ({ [path]: encodeURIComponent(value) });
  const extract = (query, path = '') => {
    const extracted = [];

    Object.keys(query).reduce((paths, key) => {
      const value = query[key];
      const current = `${path}[${key}]`;
      const isSet = Set.prototype.isPrototypeOf(value);

      if (isSet || typeof value === 'string') {
        return isSet
          ? [...value].forEach(val =>
              extracted.push(create(`${current}[]`, val)),
            )
          : extracted.push(create(current, value));
      }
      extracted.push(...extract(value, current));
    }, {});

    return extracted;
  };

  return extract(query)
    .map(path =>
      Object.entries(path).reduce((queryString, [key, value]) => {
        return `${baseName}${queryString}${key}=${value}`;
      }, ''),
    )
    .join('&');
};
```